File: 15**.cpp
// O(n^2logn)
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        int a[3];
        for(int i = 0; i < nums.size(); i++) {
            if(i && nums[i] == nums[i - 1]) continue; // duplicate
            for(int j = i + 1; j < nums.size(); j++) {
                if(j - 1 != i && nums[j] == nums[j - 1]) continue; //duplicate
                int tmp = 0 - nums[i] - nums[j];
                vector<int>::iterator it = lower_bound(nums.begin() + j + 1, nums.end(), tmp);
                if(it != nums.end()) a[2] = *it; 
                else continue;
                a[0] = nums[i], a[1] = nums[j];
                vector<int> v(a, a + 3);
                if(a[2] == tmp) ans.push_back(v);
            }
        }
        return ans;
    }
};

File: 15**.java
//O(n^2)
public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new LinkedList<>();
        Arrays.sort(nums);
        for(int i = 0; i < nums.length; i++) {
            if(i > 0 && nums[i] == nums[i - 1]) continue;
            int target = 0 - nums[i];
            
            int lo = i + 1, hi = nums.length - 1;
            while(lo < hi) {
                int sum = nums[lo] + nums[hi];
                if(sum < target) {
                    lo++;
                }
                else if(sum > target) {
                    hi--;
                } else {
                    while(lo < hi && nums[lo] == nums[lo + 1]) lo++;
                    while(lo < hi && nums[hi] == nums[hi - 1]) hi--;
                    ans.add(Arrays.asList(nums[i], nums[lo], nums[hi]));
                    lo++; hi--;
                }
            }
        }
        return ans;
    }
}
